---
title: "Eva_RNAseq Pipeline"
author: "Keshav Prasad Gubbi"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: yes
    fig_width: 10
    fig_height: 10
    fig_caption: yes
    number_sections: yes
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = TRUE, warning = FALSE)
```

```{r Package Declarations}
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("PoiClaClu"))
suppressPackageStartupMessages(library("RColorBrewer"))
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("PoiClaClu"))
suppressPackageStartupMessages(library("vsn"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library("gplots"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("genefilter"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("AnnotationDbi"))
suppressPackageStartupMessages(library("apeglm"))
suppressPackageStartupMessages(library("ComplexHeatmap"))
suppressPackageStartupMessages(library("clusterProfiler"))
suppressPackageStartupMessages(library("ggrepel"))
suppressPackageStartupMessages(library("corrplot"))
suppressPackageStartupMessages(library("GO.db"))
suppressPackageStartupMessages(library("edgeR"))
suppressPackageStartupMessages(library("GOstats"))
suppressPackageStartupMessages(library("pathview"))
suppressPackageStartupMessages(library("gage"))
suppressPackageStartupMessages(library("gageData"))
suppressPackageStartupMessages(library("GOSemSim"))
suppressPackageStartupMessages(library("DOSE"))
suppressPackageStartupMessages(library("enrichplot"))
suppressPackageStartupMessages(library("ggnewscale"))
suppressPackageStartupMessages(library("glue"))
suppressPackageStartupMessages(library("ggupset"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("stats"))
suppressPackageStartupMessages(library("FactoMineR"))
suppressPackageStartupMessages(library("factoextra"))
suppressPackageStartupMessages(library("pcaExplorer"))
suppressPackageStartupMessages(library("here"))
```

```{r File Path Declarations }
here::i_am("Eva_TotalRNASeq.rmd")
here()

# Project Path
Eva_project_path <- here()

# Provide the countsmatrix and the samples
countsmatrix_path <- file.path(Eva_project_path, "featurecounts_Eva_totalRNAseq.csv")
samples_path <- file.path(Eva_project_path, "Samples_eva.csv")

# Folder Paths for Different Comparisons
# adult_spfVsd7_spf_path <- file.path(here(), "adult_spfVsd7_spf")
# d7_GFVsd7_spF_path <- file.path(here(), "d7_GFVsd7_spF")
adult_GFVsd7_GF_path <- file.path(here(), "adult_GFVsd7_GF")
# Day5_UninfectedVSInfected_path <- file.path(here(), "Day5_UninfectedVSInfected")
# Day_3UninfectedVS5Uninfected_path <- file.path(here(), "Day_3UninfectedVS5Uninfected")
# Day_3InfectedVS5Infected_path <- file.path(here(), "Day_3InfectedVS5Infected")

# for UninfectedVSInfected is the volcano plot notation: for Deseq2 results, the infected is Nr.
# and uninfected is Dr.
```
```{r}
# Comparison_Path! NEEDS TOBE UPDATED FOR EVERY NEW COMPARISON BEING RUN for pipeline!
# Comparison_Path <- adult_spfVsd7_spf_path
Comparison_Path <- adult_GFVsd7_GF_path
#setwd(Comparison_Path)
# Also Create a comparison Variable: That Could be used later for all other comparison titles using a glue Variable.
# Comparison <- "Day3_Uninfected VS Infected"
Comparison <- "adult GermFree Vs Day7 GermFree"

# Paths for Different Plots and Tables for each comparison. To be updated for each respective 
# comparison that is being run for
SignificantGenes_path <- file.path(Comparison_Path, glue( "/SignificantGenes_{Comparison}.csv"))
results_path <- file.path(Comparison_Path, glue("/results_DGE_{Comparison}.csv"))
UpGeneList_path <- file.path(Comparison_Path, glue("/UpGeneList_DGE_{Comparison}.csv"))
DownGeneList_path <- file.path(Comparison_Path, glue("/DownGeneList_DGE_{Comparison}.csv"))
GO_UPReg_path <- file.path(Comparison_Path, glue("/GO_UPReg_results_{Comparison}.csv"))
GO_UPReg_TCell_path <- file.path(Comparison_Path, glue("/GOTerms_Tcell_{Comparison}.csv"))
GO_DOWNReg_path <- file.path(Comparison_Path, glue("/GO_DOWNReg_results_{Comparison}.csv"))
termSim_values_path <- file.path(Comparison_Path, glue("/termSim_values_{Comparison}.csv"))
```
```{r counts Matrix processing}
## Input the Count Matrix

countsmatrix <- read.csv(countsmatrix_path)

### Clean up the Count Matrix
rownames(countsmatrix) <- countsmatrix[, 2]
# converting first column of gene ID into rownames, to be used for sanity check later
```
```{r GenderGenes Filtering}
gendergenes_biomart <- data.frame(read.csv("~/R/Eva_TotalRNASeq/20221017_y-chromosomal genes_biomart.txt", stringsAsFactors = TRUE))

nrow(gendergenes_biomart)

#na.omit(gendergenes_biomart)

## Remove rows with only empty cells
gendergenes_biomart <- gendergenes_biomart[!apply(gendergenes_biomart == " ", 1, all), ]                
nrow(gendergenes_biomart)

## Removal of Gender Genes from ENSEMBL ID column itself
## Use %in% to identify which elements from the first data frame are not contained in the second data frame, then pass the resulting logical vector to the first data frame to subset.
countsmatrix <- countsmatrix[(!countsmatrix$EnsemblID %in% gendergenes_biomart$Gene.stable.ID), ,
             drop = FALSE]


countsmatrix <- countsmatrix %>% filter(
  countsmatrix$EnsemblID != "ENSMUSG00000086503",
  countsmatrix$EnsemblID != "ENSMUSG00000097571",
  countsmatrix$EnsemblID != "ENSMUSG00000086370",
  countsmatrix$EnsemblID != "ENSMUSG00000031329",
  countsmatrix$EnsemblID != "ENSMUSG00000030057"
)
nrow(countsmatrix)
```

```{r}
# It is IMPORTANT to keep the names of the genes in the rownames
countsmatrix <- subset(countsmatrix, select = -c(X, EnsemblID)) # dropping the X column

## Display the column names
# colnames(countsmatrix)

### Annotating and Exporting ENSEMBL ID into Gene Symbols

# Adding genes annotated from ENSEMBL ID to Gene symbols and ENTREZ Id to countsmatrix table.
# Will be keeping the symbols and entrez columsn to be added later into results table as it is for later use

cm_row <- rownames(countsmatrix)
head(cm_row)

# Mapping the ENSEMBL ID to Symbol and ENTREZ ID
symbols <- mapIds(
  org.Mm.eg.db,
  keys = cm_row,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

# Create symbols column
symbols <- symbols[!is.na(symbols)]
symbols <- symbols[match(rownames(countsmatrix), names(symbols))]

# Creating a new column called genename and putting in the symbols and entrez columns into count matrix
countsmatrix$genename <- symbols

# Removing all rows with NA values for genenames, so that those rows are filtered out.
countsmatrix <- unique(countsmatrix[rowSums(is.na(countsmatrix)) == 0, ]) # Apply rowSums & is.na

nrow(countsmatrix)

# Moving the ENSEMBL ID from rownames into separate column for itself.
countsmatrix <- tibble::rownames_to_column(countsmatrix, "E_ID")
# Removing the duplicated genes so that then these genes can be made into rownames for countsmatrix
countsmatrix <- distinct(countsmatrix[!duplicated(countsmatrix$genename), ])

# Now make the ganename column into rownames of count matrix
rownames(countsmatrix) <- countsmatrix[, "genename"]

# Dropping the column E_ID, genenames so that only numeric values are present in it as an input of DESEq Object.
countsmatrix <- subset(countsmatrix, select = -c(genename, E_ID))

# Changing countsmatrix into Matrix of numeric values so that only numeric values are present in it as an input of DESEq Object.
countsmatrix <- as.matrix(countsmatrix)
class(countsmatrix) <- "numeric"

### The Count Matrix is:
head(countsmatrix, 2)
```

```{r coldata processing}
## Develop ColData by importing metadata

# Read the csv file and change the column name. the samples.csv is a list of sample names, ie, the names of bam files.
sample_ID <- read.csv(samples_path)

coldata <- as.data.frame(sample_ID)
colnames(coldata) <- c("Sample_Name", "MouseType", "condition") # change name of one of the columns
rownames(coldata) <- coldata[, 1] # move the sample names into row names
# The metadata can be found in a df called coldata!
head(coldata)

## the elements from Sample_Name from coldata must the the colnames of countsmatrix
colnames(countsmatrix) <- coldata$Sample_Name
```
### Reduce larger Matrix to smaller one - based on comparison
```{r The Processed Input is Ready!}

# *****************Now to the comparisons*************
# d7 spF vs adult spF
# coldata <- coldata[c(9, 10, 11, 18, 19, 20), ]
# countsmatrix <- countsmatrix[, c(9, 10, 11, 18, 19, 20)]
# ******************************************************
# d7 GF vs adult spF
# coldata <- coldata[c(5, 6, 7, 8, 9, 10, 11), ]
# countsmatrix <- countsmatrix[, c(5, 6, 7, 8, 9, 10, 11)]

# d7 GF vs adult spF
coldata <- coldata[c(5, 6, 7, 8, 15, 16, 17), ]
countsmatrix <- countsmatrix[, c(5, 6, 7, 8, 15, 16, 17)]
```

```{r Save Plot}
# Function to save generic plots
saveplot <- function(plot, name) {
  # Function to save the plots
  ggsave(filename = file.path(Comparison_Path, glue("/{name}_{Comparison}.png")),
        plot = plot,
        dpi = 300,
        width = 25,
        height = 25,
        units = "in")
  dev.off()
}
```
```{r DESeq Analysis}

# Sanity Check for DDS
all(rownames(coldata) %in% colnames(countsmatrix))
ncol(countsmatrix) == nrow(coldata)
dim(countsmatrix)


## Creating the DESeq Data set Object
dds <- DESeqDataSetFromMatrix(
  countData = countsmatrix,
  colData = coldata,
  #design = ~MouseType
  design = ~condition
)
```

```{r}
keep <- rowSums(counts(dds)) > 10
dds <- dds[keep, ]
nrow(dds)
## Applying VST transformation
vsd <- vst(dds, blind = FALSE)
# head(assay(vsd), 3)
colData(vsd)
vsd_coldata <- colData(vsd)
dds <- estimateSizeFactors(dds)
### Euclidean Distance between samples
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$Sample_Name
colnames(sampleDistMatrix) <- vsd$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(EuclideanDistanceHeatmap <- pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  main = glue("Sample-to-Sample Euclidean Distance of Samples: {Comparison}"),
  col = colors
))
### Poisson Distance between Samples
poisd <- PoissonDistance(t(counts(dds))) # raw counts or unnormalised data
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- dds$Sample_Name
colnames(samplePoisDistMatrix) <- dds$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(poisson_dist_plot <- pheatmap(samplePoisDistMatrix,
  clustering_distance_rows = poisd$dd,
  clustering_distance_cols = poisd$dd,
  main = glue("Sample-to-Sample Poisson Distance of Samples: {Comparison}"),
  col = colors
))

#savingFunction(poisson_dist_plot, "poisson_dist_plot")
```
## PCA Plot
```{r PCA Plot Aesthetics}
# PCA Plot
### Functions for Plot aethetics and saving PCA Plots
color_values <- c("black", "black", "red", "red", "red", "red", "black")
# the basic set of common aesthetic settings for PCA plots,
theme.my.own <- list(theme_bw(),
                      geom_point(size = 3),
                      coord_fixed(),
                      scale_y_continuous(breaks = seq(-100, 100, 10), 
                                         sec.axis = sec_axis(~ . * 1, labels = NULL, breaks = NULL)),
                      scale_x_continuous(breaks = seq(-100, 100, 10), 
                                         sec.axis = sec_axis(~ . * 1, labels = NULL, breaks = NULL)),
                      theme_classic(),
                      geom_hline(yintercept = 0, color = "gray", size = 1),
                      geom_vline(xintercept = 0, color = "gray", size = 1),
                      theme(text = element_text(size = 15),
                            axis.text = element_text(size = 15),
                            legend.position = "bottom",
                            aspect.ratio = 1),
                      # geom_text(size = 4, hjust = 0, vjust = 0)
                      geom_text_repel(size = 5, min.segment.length = 0.5)
)
```

```{r PCA Plot Calculation}
## Calculating all PCA Values
plotPCA_local <- function(object,
                          intgroup = "condition",
                          ntop = 500,
                          returnData = TRUE,
                          nPC = 4) {
  # calculate the variance for each gene
  rv <- rowVars(assay(object))
  ntop <- 500
  # select the ntop genes by variance
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select, ]))
  # summary(pca)
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <-
    as.data.frame(colData(object)[, intgroup, drop = FALSE])
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = ":"))
  } else {
    colData(object)[[intgroup]]
  }
  # assembly the data for the plot
  d <- cbind(
    pca$x[, seq_len(min(nPC, ncol(pca$x))), drop = FALSE],
    data.frame(group = group, intgroup.df, name = colnames(object))
  )
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:nPC]
    # l <- list(pca,d)
    # return(l)
    return(d)
  }
}
## PCA Plot with VST Data
### Function for calculating percentvar
percentvar_calculation <- function(pcaData_variable) {
  # function to calculate percentvar for different variables
  percentvar_variable <- round(100 * attr(pcaData_variable, "percentVar"), digits = 3)
  return(percentvar_variable)
}

savingFunction <- function(plotname, metadatacolumn) {
  # Function to save the PCA plots
  ggsave(filename = glue(Comparison_Path, "/PCAplot_{metadatacolumn}.png"),
          plot = plotname,
          dpi = 300,
          width = 10,
          height = 10,
          units = "in"
  )
}

pcaData <- plotPCA_local(vsd, intgroup = c("condition", "Sample_Name"), returnData = T)
#pcaData <- plotPCA_local(vsd, intgroup = c("MouseType", "Sample_Name"), returnData = T)
pcaData
percentVar <- percentvar_calculation(pcaData)
percentVar
```

```{r PC Plot: PC1 vs PC2, fig.height=8, fig.width=8}
(PCAplot_vst <- ggplot(pcaData, aes(x = PC1, y = PC2, color = Sample_Name, label = Sample_Name)) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggtitle(glue("PCA: {Comparison}")) +
  scale_colour_manual(values = color_values) +
  theme.my.own)
savingFunction(PCAplot_vst, metadatacolumn = glue("PC1vsPC2_{Comparison}"))

```
```{r PC Plot: PC3 vs PC4, fig.height=8, fig.width=8}
(PCAplot_pc34 <- ggplot(
  pcaData,
  aes(x = PC3,y = PC4, color = Sample_Name, label = Sample_Name)) +
  xlab(paste0("PC3: ", percentVar[3], "% variance")) +
  ylab(paste0("PC4: ", percentVar[4], "% variance")) +
  ggtitle(glue("PCA: {Comparison}")) +
  scale_colour_manual(values = color_values) +
  theme.my.own)
savingFunction(PCAplot_pc34, metadatacolumn = glue("PC3vsPC4_{Comparison}"))
```
```{r FactoExtra}
# calculate the variance for top 500 gene
rv <- rowVars(assay(vsd))
ntop <- 500
# select the ntop genes by variance
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
df1 <- t(assay(vsd)[select, ])

res.pca <- PCA(df1, graph = FALSE, scale.unit = FALSE)
summary.PCA(res.pca)

# Visualize eigenvalues/variances
fviz_screeplot(res.pca, addlabels = TRUE)

eig.val <- get_eigenvalue(res.pca)
var <- get_pca_var(res.pca)
# fviz_pca_var(res.pca, repel = TRUE)
```
```{r BiPlots, fig.width=8, fig.height=8}
## Genes + PCA Biplots
heat.colors <- brewer.pal(6, "RdYlBu")

## Genes + PCA Biplots
(Genes_Biplot <- fviz_pca_biplot(res.pca, repel = TRUE))
savingFunction(Genes_Biplot, "Genes_Biplot")

(Genes_contributions_Biplot <- fviz_pca_var(res.pca,
  col.var = "contrib", repel = TRUE,
  gradient.cols = c("Gray", "blue", "pink", "yellow", "orange", "green", "red", "black"),
))
savingFunction(Genes_contributions_Biplot, "Genes_contributions_Biplot")

# Contributions of variables to PC2

(top25_genes_dim2 <- fviz_contrib(res.pca, choice = "var", axes = 2, top = 25))
savingFunction(top25_genes_dim2, "top25_genes_dim2")
# # Contributions of variables to PC1
(top25_genes_dim1 <- fviz_contrib(res.pca, choice = "var", axes = 1, top = 25))
savingFunction(top25_genes_dim1, "top25_genes_dim1")
```
```{r}
## Hierarchical Clustering
### applying rlog Transformation
rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)
### Extract the rlog matrix from the object
rld_mat <- assay(rld) # assay() is function from the "SummarizedExperiment" package that was loaded when you loaded DESeq2
### Compute pairwise correlation values
rld_cor <- cor(rld_mat) ## cor() is a base R function
head(rld_cor) ## check the output of cor(), make note of the rownames and colnames
### Plot heatmap
heat.colors <- brewer.pal(6, "RdYlBu")
(Hclust_plot <- pheatmap(rld_cor,
  color = heat.colors,
  main = glue("Heirarchical Clustering of Samples - {Comparison}")
  # filename = 'C:/Users/kesha/Documents/Eva_TotalRNASeq/Hclust_plot.tiff'
))
# Hclust_plot
```

```{r DGE Results}
# DGE Results
### Running the differential expression pipeline
dds1 <- DESeq(dds)
### Building the results table
res <- results(dds1, contrast = c("condition", "d7", "adult"))
#res <- results(dds1, contrast = c("MouseType", "GF", "SPF"))
resdf <- as.data.frame(res) # convert the results table to a df
head(resdf, 20)
write.csv(resdf, file = results_path)
```

### Histogram of p-values

```{r}
hist(res$pvalue, breaks = 50, col = "grey50", border = "blue")
```
```{r}
resdf <- tibble::rownames_to_column(resdf, "symbol")
gn <- resdf$symbol
# Mapping the Symbol to ENTREZ ID
entrez <- mapIds(
  org.Mm.eg.db,
  keys = gn,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)
ensembl_id <- mapIds(
  org.Mm.eg.db,
  keys = gn,
  column = "ENSEMBL",
  keytype = "SYMBOL",
  multiVals = "first"
)
resdf$entrez <- entrez
resdf$ensemblID <- ensembl_id
```

Omit NA values from symbol and respective rows!

```{r}
resdf <- resdf %>% filter(!is.na(symbol) & !is.na(entrez))
nrow(resdf)
```
# Volcano Plots

## Volcano Plots based on Enhanced Volcano

```{r fig.height= 10, fig.width= 10}
p1 <- EnhancedVolcano(resdf,
  lab = resdf$symbol,
  x = "log2FoldChange",
  y = "pvalue",
  xlab = bquote(~ Log[2] ~ "FoldChange"),
  pCutoff = 0.05,
  FCcutoff = 1.0,
  title = glue("DE genes: Log2FoldChange Vs -Log10 Pvalue: {Comparison}"),
  subtitle = bquote(~ Log[2] ~ "|FoldChange| = 1, pvalue < 0.05"),
  pointSize = 2.0,
  labSize = 5.5,
  boxedLabels = FALSE,
  gridlines.major = FALSE,
  gridlines.minor = FALSE,
  colAlpha = 0.5,
  legendPosition = "bottom",
  legendLabSize = 12,
  legendIconSize = 4.0,
  drawConnectors = T,
  widthConnectors = 0.75,
  max.overlaps = 10,
  axisLabSize = 22
)
(volcano1 <- p1 + scale_y_continuous(
  limits = c(0, 8),
  breaks = seq(0, 8, 1),
  sec.axis = sec_axis(~ . * 1,
    labels = NULL,
    breaks = NULL
  )
) +
  scale_x_continuous(
    limits = c(-4, 5),
    breaks = seq(-4, 5, 1),
    sec.axis = sec_axis(~ . * 1,
      labels = NULL,
      breaks = NULL
    )
  ) +
  xlab(expression(DownRegulated %<->% UpRegulated))
)
ggsave(
  filename = file.path(Comparison_Path, glue("/V1_L2fcVsPvalue_{Comparison}.png")),
  plot = volcano1,
  dpi = 300,
  width = 5,
  height = 5,
  units = "in"
)
```
## volcano Plot of Log2FC vs Padj

```{r fig.height= 10, fig.width= 10}
p2 <- EnhancedVolcano(resdf,
  lab = resdf$symbol,
  x = "log2FoldChange",
  y = "padj",
  xlab = bquote(~ Log[2] ~ "FoldChange"),
  ylab = bquote(~ Log[10] ~ "Padj"),
  pCutoff = 0.05,
  FCcutoff = 1.0,
  title = glue("DE genes: Log2FoldChange Vs -Log10 Padj: {Comparison}"),
  subtitle = bquote(~ Log[2] ~ "|FoldChange| = 1, pvalue < 0.05"),
  pointSize = 2.0,
  labSize = 5.0,
  boxedLabels = FALSE,
  gridlines.major = FALSE,
  gridlines.minor = FALSE,
  colAlpha = 0.5,
  legendPosition = "bottom",
  legendLabSize = 12,
  legendIconSize = 4.0,
  drawConnectors = T,
  widthConnectors = 0.75,
  max.overlaps = 10,
  axisLabSize = 22
)
(volcano2 <- p2 + scale_y_continuous(
  limits = c(0, 4),
  breaks = seq(0, 4, 1),
  sec.axis = sec_axis(~ . * 1,
    labels = NULL,
    breaks = NULL
  )
) +
  scale_x_continuous(
    limits = c(-3, 4),
    breaks = seq(-3, 4, 1),
    sec.axis = sec_axis(~ . * 1,
      labels = NULL,
      breaks = NULL
    )
  ) +
  xlab(expression(DownRegulated %<->% UpRegulated))
)
ggsave(
  filename =  file.path(Comparison_Path, glue("/V2_L2fcVsPadj_{Comparison}.png")),
  plot = volcano2,
  dpi = 300,
  width = 10,
  height = 10,
  units = "in"
)
```

## Significant Differentially Expressed Genes

```{r}
resdf$diffexpressed <- "NS"
# if log2Foldchange > 1.0 and pvalue < 0.05, set as "UP"
resdf$diffexpressed[resdf$log2FoldChange > 1.0 & resdf$pvalue < 0.05] <- "UP"
# if log2Foldchange < -1.0 and pvalue < 0.05, set as "DOWN"
resdf$diffexpressed[resdf$log2FoldChange < -1.0 & resdf$pvalue < 0.05] <- "DOWN"
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
resdf$delabel <- NA
resdf$delabel[resdf$diffexpressed != "NS"] <- resdf$symbol[resdf$diffexpressed != "NS"]
```

### Arrive at relevant genes by imposing thresholds.

```{r}
sigsdf <- resdf[(abs(resdf$log2FoldChange) > 1) & (resdf$pvalue < 0.05), ]
nrow(sigsdf)
write.csv(sigsdf, SignificantGenes_path)
```
### Gender Genes Removed from Table
```{r Verify if any Gender Genes still left}
genedergenes_df <- sigsdf[is.element(sigsdf$symbol, c("Xist","Jpx","Ftx","Tsx","Cnbp2")), ]
genedergenes_df
```
Therefore, gender genes werent so much in play in terms of significance!

### Number of Genes from different strains that are contributing to UP/DOWN regulation.

```{r}
sigsdf_UP <- sigsdf[(sigsdf$log2FoldChange) > 1, ] # UP Regulation Table
sigsdf_DOWN <- sigsdf[(sigsdf$log2FoldChange) < -1, ] # DOWN Regulation Table
nrow(sigsdf_UP)
nrow(sigsdf_DOWN)
```
```{r}
# sorted based on highest Log2FC value
UpGene <- sigsdf_UP[order(-sigsdf_UP$log2FoldChange, decreasing = TRUE), ]$symbol
DownGene <- sigsdf_DOWN[order(-sigsdf_DOWN$log2FoldChange, decreasing = TRUE), ]$symbol
```
```{r}
write.csv(UpGene, UpGeneList_path)
write.csv(DownGene, DownGeneList_path)
```

### Top20 UP Genes and Top20 DOWN Genes

#### UpGene
```{r}
head(UpGene, 20)
```

#### DownGene
```{r}
head(DownGene, 20)
```
## Z-score based Gene Heatmaps

```{r}
# Determining the significant Genes based on Log2FC and pvalue thresholds
sigs2df <- resdf[(abs(resdf$log2FoldChange) > 1) & (resdf$pvalue < 0.05), ]
```
### with Whole table (all genes together!)

```{r}
# mat <- counts(dds1, normalized = TRUE)[rownames(sigsdf),]
mat <- counts(dds1, normalized = TRUE)[(sigsdf$symbol) %in% rownames(counts(dds1)), ]
mat.zs <- t(apply(mat, 1, scale)) # Calculating the zscore for each row
colnames(mat.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat.zs)
```

```{r}
mat2 <- counts(dds, normalized = TRUE)[(sigsdf$symbol) %in% rownames(counts(dds)), ]
mat2.zs <- t(apply(mat2, 1, scale)) # Calculating the zscore for each row
colnames(mat2.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat2.zs)
# # pheatmap(mat2.zs, cluster_cols = TRUE, cluster_rows = FALSE, show_rownames = FALSE)
```
```{r fig.height=8, fig.width=8}
AllGenes_Heatmap <- Heatmap(mat2.zs,
                              cluster_columns = TRUE,
                              cluster_rows = TRUE,
                              column_labels = colnames(mat2.zs),
                              name = glue("DE Genes- {Comparison}"),
                              show_row_names = FALSE,
                              use_raster = TRUE,
                              raster_quality = 5,
                              column_names_gp = grid::gpar(fontsize = 12),
                              #row_labels = sigs2df[rownames(mat2.zs), ]$symbol
                               heatmap_legend_param = list(legend_direction = "horizontal", 
                                                           legend_width = unit(5, "cm"))
                             )
png(file = file.path(Comparison_Path, glue("/DEGenes_heatmap1_{Comparison}.png") ),
    width = 1000, height = 1000, units = "px", pointsize = 12,
    bg = "white", res = NA, family = "", restoreConsole = TRUE,
    type = "windows",
    symbolfamily = "default"
    )
draw(AllGenes_Heatmap, heatmap_legend_side = "bottom")
dev.off()
#savingFunction(AllGenes_Heatmap, "AllGenes_Heatmap")
```
```{r fig.height= 500, fig.width= 10}
# LongHeatMap_Allgenes <- Heatmap(mat2.zs,
#                                 cluster_columns = TRUE,
#                                 cluster_rows = TRUE,
#                                 column_labels = colnames(mat2.zs),
#                                 row_labels = rownames(mat2.zs),#sigsdf[rownames(mat2.zs), ]$symbol
#                                 name = "All Genes",
#                                 show_row_names = TRUE,
#                                 use_raster = TRUE,
#                                 raster_quality = 5,
#                                 column_names_gp = grid::gpar(fontsize = 12),
#                                 #row_labels = sigsdf[rownames(mat2.zs), ]$symbol
# )
# png(file = file.path(Comparison_Path, glue("/DEGenes_Longheatmap_{Comparison}.png") ),
#     width = 800, height = 800, units = "px", pointsize = 12,
#     bg = "white", res = NA, family = "", restoreConsole = TRUE,
#     type = "windows",
#     symbolfamily = "default"
#     )
# draw(LongHeatMap_Allgenes, heatmap_legend_side = "bottom")
# dev.off()
```

Need to filter these results to accommodate better the heat maps and also volcano plots

### with tighter constraints (all genes together!)

```{r}
sigs1df <- resdf[(resdf$baseMean > 15000) & (abs(resdf$log2FoldChange) > 2) & (resdf$pvalue < 0.05), ]
mat1 <- counts(dds, normalized = TRUE)[(sigs1df$symbol), ]
mat1.zs <- t(apply(mat1, 1, scale)) # Calculating the zscore for each row
colnames(mat1.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat1.zs)
```

```{r fig.height=12}
Tightconstraints_Heatmap <- Heatmap(mat1.zs,
                                    cluster_columns = TRUE,
                                    cluster_rows = TRUE,
                                    column_labels = colnames(mat1.zs),
                                    name = glue("DE Genes - {Comparison}"),
                                    row_labels = rownames(mat1.zs),
                                    column_names_gp = grid::gpar(fontsize = 18),
                                    row_names_gp = grid::gpar(fontsize = 20),
                                    heatmap_legend_param = list(legend_direction = "horizontal", 
                                                                legend_width = unit(5, "cm")
                                                                )
                                    )
png(file = file.path(Comparison_Path, glue("/DEGenes_heatmap2_{Comparison}.png") ),
    width = 1000, height = 1000, units = "px", pointsize = 12,
    bg = "white", res = NA, family = "", restoreConsole = TRUE,
    type = "windows",
    symbolfamily = "default"
    )
draw(Tightconstraints_Heatmap, heatmap_legend_side = "bottom")
dev.off()
```
# GO Terms with clusterProfiler

## GO Terms for UP Regulated Genes

### GO over-representation analysis for UP Regulated Genes
```{r}
UPgene_ENS_ID <- (sigsdf_UP$ensemblID)
GO_UPRegResults <- enrichGO(
  gene = UPgene_ENS_ID,
  OrgDb = "org.Mm.eg.db",
  keyType = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)
```

```{r}
GO_UpRegdf <- as.data.frame(GO_UPRegResults)
head(GO_UpRegdf, 20)
write.csv(GO_UpRegdf, GO_UPReg_path)
```
```{r T-Cell Based GO Terms}
# create a Cytokine based GO term Data frame by extracting the rows that at least partially matches to the word Cytokine
GO_UpRegdf_TCell <- GO_UpRegdf[str_detect(GO_UpRegdf$Description, "T cell"), ]
#Warning: I wont be able to detect GO terms that do not have the word cytokines mentioned in their description.
#
head(GO_UpRegdf_TCell, 20)
write.csv(GO_UpRegdf_TCell, GO_UPReg_TCell_path)
```
```{r fig.height=12}
GO_UPReg_Barplot <- plot(barplot(GO_UPRegResults,
  showCategory = 20,
  font.size = 25,
  title = " Up Regulated",
  label_format = 50
))
saveplot(GO_UPReg_Barplot, "GO_UPReg_Barplot")
```

```{r fig.height=12}
GO_UPReg_Dotplot <- plot(dotplot(GO_UPRegResults,
  showCategory = 20,
  font.size = 30,
  title = " Up Regulated",
  label_format = 50
))
# GO_UPReg_Dotplot1 <- GO_UPReg_Dotplot + scale_y_discrete(labels=function(y) str_wrap(y, width=40))
saveplot(GO_UPReg_Dotplot, "GO_UPReg_Dotplot")
```

```{r fig.height=10, fig.width=12}
GO_UPReg_Cnetplot <- plot(cnetplot(GO_UPRegResults,
  showCategory = 15,
  font.size = 25,
  label_format = 75
))
saveplot(GO_UPReg_Cnetplot, "GO_UPReg_Cnetplot")
```

### Upset Plot

```{r fig.height= 8, fig.width=12}
GO_UPReg_Upsetplot <- plot(upsetplot(GO_UPRegResults))
saveplot(GO_UPReg_Upsetplot, "GO_UPReg_Upsetplot")
```

### Heatplot

The heatplot is similar to cnetplot, while displaying the relationships as a heatmap. The gene-concept network may become too complicated if user want to show a large number significant terms. The heatplot can simplify the result and more easy to identify expression patterns.

```{r fig.height= 12, fig.width=15}
GO_UPReg_Heatplot <- plot(heatplot(GO_UPRegResults, label_format = 150))
saveplot(GO_UPReg_Heatplot, "GO_UPReg_Heatplot")
```

### Tree Plot of Enriched Terms

```{r fig.height= 10, fig.width=15}
edox2 <- pairwise_termsim(GO_UPRegResults)
GO_UPReg_enrichtreeplot <- plot(treeplot(edox2))
saveplot(GO_UPReg_enrichtreeplot, "GO_UPReg_enrichtreeplot")
```
```{r}
termSim_values <- tibble(edox2@termsim)
termSim_values
write.csv(termSim_values,termSim_values_path)
```

```{r fig.height=12, fig.width=12}
GO_UPReg_emapplot <- emapplot(edox2, showCategory = 25, repel = TRUE)
#saveplot(GO_UPReg_emapplot, "GO_UPReg_emapplot")
#dev.off()
```
## GO Terms for Down Regulated Genes

### GO over-representation analysis for DOWN Regulated Genes

```{r}
DOWNgene_ENS_ID <- (sigsdf_DOWN$ensemblID)
GO_DOWNRegResults <- enrichGO(
  gene = DOWNgene_ENS_ID,
  OrgDb = "org.Mm.eg.db",
  keyType = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)
```

```{r}
GO_DOWNRegdf <- as.data.frame(GO_DOWNRegResults)
head(GO_DOWNRegdf)
write.csv(GO_DOWNRegdf, GO_DOWNReg_path)
```

```{r fig.height=10}
(GO_DOWNReg_Barplot <- plot(barplot(GO_DOWNRegResults,
  showCategory = 25,
  font.size = 15,
  title = glue("Down Regulated_{Comparison}"),
  label_format = 40
  )))
saveplot(GO_DOWNReg_Barplot, "GO_DOWNReg_Barplot")
```

```{r fig.height=10}
GO_UPReg_Dotplot <- plot(dotplot(GO_DOWNRegResults,
  showCategory = 25,
  font.size = 15,
  title = glue("Down Regulated_{Comparison}"),
  label_format = 40
  ))

#saveplot(GO_DOWNReg_Dotplot, "GO_DOWNReg_Dotplot")
```

```{r fig.height=10, fig.width=10}
(GO_DOWNReg_Cnetplot <- plot(cnetplot(GO_DOWNRegResults,
  showCategory = 10,
  circular = FALSE,
  # colorEdge = TRUE
)))

saveplot(GO_DOWNReg_Cnetplot, "GO_DOWNReg_Cnetplot")
```

### Upset Plot

The upsetplot is an alternative to cnetplot for visualizing the complex
association between genes and gene sets. It emphasizes the gene
overlapping among different gene sets.

```{r}
GO_DOWNReg_Upsetplot <- plot(upsetplot(GO_DOWNRegResults))
saveplot(GO_DOWNReg_Upsetplot, "GO_DOWNReg_Upsetplot")
```

### Heat Plot

```{r}
GO_DOWNReg_Heatplot <- plot(heatplot(GO_DOWNRegResults))
saveplot(GO_DOWNReg_Heatplot, "GO_DOWNReg_Heatplot")
```

### Tree Plot of Enriched Terms

```{r}
edox1 <- pairwise_termsim(GO_DOWNRegResults)
GO_DOWNReg_enrichtreeplot <- plot(treeplot(edox1))
saveplot(GO_DOWNReg_enrichtreeplot, "GO_DOWNReg_enrichtreeplot")
```
