---
title: "Eva_RNAseq - d7 spF vs adult spF"
author: "Keshav Prasad Gubbi"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: yes
    fig_width: 10
    fig_height: 12
    fig_caption: yes
    number_sections: yes
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages(library("dplyr"))

suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("PoiClaClu"))
suppressPackageStartupMessages(library("RColorBrewer"))
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("PoiClaClu"))
suppressPackageStartupMessages(library("vsn"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library("gplots"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("genefilter"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("AnnotationDbi"))
suppressPackageStartupMessages(library("apeglm"))
suppressPackageStartupMessages(library("ComplexHeatmap"))
suppressPackageStartupMessages(library("clusterProfiler"))
suppressPackageStartupMessages(library("ggrepel"))
suppressPackageStartupMessages(library("corrplot"))
suppressPackageStartupMessages(library("GO.db"))
suppressPackageStartupMessages(library("edgeR"))
suppressPackageStartupMessages(library("GOstats"))
suppressPackageStartupMessages(library("pathview"))
suppressPackageStartupMessages(library("gage"))
suppressPackageStartupMessages(library("gageData"))
suppressPackageStartupMessages(library("GOSemSim"))
suppressPackageStartupMessages(library("DOSE"))
suppressPackageStartupMessages(library("enrichplot"))
suppressPackageStartupMessages(library("ggnewscale"))
suppressPackageStartupMessages(library("glue"))
suppressPackageStartupMessages(library("ggupset"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("stats"))
suppressPackageStartupMessages(library("FactoMineR"))
suppressPackageStartupMessages(library("factoextra"))
suppressPackageStartupMessages(library("pcaExplorer"))
```

```{r}
## Input the Count Matrix

countsmatrix <- read.csv("C:/Users/kesha/Documents/Eva_TotalRNASeq/featurecounts_Eva_totalRNAseq.csv",
                         stringsAsFactors = TRUE)


### Clean up the Count Matrix
rownames(countsmatrix) <- countsmatrix[, 2]
# converting first column of gene ID into rownames, to be used for sanity check later

# It is IMPORTANT to keep the names of the genes in the rownames
countsmatrix <- subset(countsmatrix, select = -c(X, EnsemblID)) # dropping the X column

## Display the column names
#colnames(countsmatrix)

### Annotating and Exporting ENSEMBL ID into Gene Symbols

# Adding genes annotated from ENSEMBL ID to Gene symbols and ENTREZ Id to countsmatrix table.
# Will be keeping the symbols and entrez columsn to be added later into results table as it is for later use

cm_row <- rownames(countsmatrix)
head(cm_row)

# Mapping the ENSEMBL ID to Symbol and ENTREZ ID
symbols <- mapIds(
  org.Mm.eg.db,
  keys = cm_row,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

# Create symbols column
symbols <- symbols[!is.na(symbols)]
symbols <- symbols[match(rownames(countsmatrix), names(symbols))]

# Creating a new column called genename and putting in the symbols and entrez columns into count matrix
countsmatrix$genename <- symbols

# Removing all rows with NA values for genenames, so that those rows are filtered out.
countsmatrix <- unique(countsmatrix[rowSums(is.na(countsmatrix)) == 0,]) # Apply rowSums & is.na

nrow(countsmatrix)

# Moving the ENSEMBL ID from rownames into separate column for itself.
countsmatrix <- tibble::rownames_to_column(countsmatrix, "E_ID")
# Removing the duplicated genes so that then these genes can be made into rownames for countsmatrix
countsmatrix <- distinct(countsmatrix[!duplicated(countsmatrix$genename),])

# Now make the ganename column into rownames of count matrix
rownames(countsmatrix) <- countsmatrix[, "genename"]

# Dropping the column E_ID, genenames so that only numeric values are present in it as an input of DESEq Object.
countsmatrix <- subset(countsmatrix, select = -c(genename, E_ID))

# Changing countsmatrix into Matrix of numeric values so that only numeric values are present in it as an input of DESEq Object.
countsmatrix <- as.matrix(countsmatrix)
class(countsmatrix) <- "numeric"

### The Count Matrix is:
head(countsmatrix, 2)

```

```{r}
## Develop ColData by importing metadata

# Read the csv file and change the column name. the samples.csv is a list of sample names, ie, the names of bam files.
sample_ID <- read.csv( "C:/Users/kesha/Documents/Eva_TotalRNASeq/Samples_eva.csv")

coldata <- as.data.frame(sample_ID)
colnames(coldata) <- c("Sample_Name", "MouseType", "condition") # change name of one of the columns
rownames(coldata) <- coldata[, 1] # move the sample names into row names
# The metadata can be found in a df called coldata!
head(coldata)

## the elements from Sample_Name from coldata must the the colnames of countsmatrix
colnames(countsmatrix) <- coldata$Sample_Name
```

```{r}

# *****************Now to the comparisons*************
# d7 spF vs adult spF
coldata <- coldata[c(9,10,11,18,19,20),]
countsmatrix <- countsmatrix[ ,c(9,10,11,18,19,20)]
# ******************************************************
```
```{r}

# Function to save generic plots
saveplot <- function(plot, name) {
  # Function to save the plots
  ggsave(
    filename =
      glue("C:/Users/kesha/Documents/Eva_TotalRNASeq/{name}.png"),
    plot = plot,
    dpi = 300,
    width = 25,
    height = 25,
    units = "in"
  )
}

```

```{r}
# Sanity Check for DDS
all(rownames(coldata) %in% colnames(countsmatrix))
ncol(countsmatrix) == nrow(coldata)
dim(countsmatrix)
```

```{r}
## Creating the DESeq Data set Object
dds_infected <- DESeqDataSetFromMatrix(
  countData = countsmatrix,
  colData = coldata,
  design = ~condition
)
nrow(dds_infected)

```

```{r}
keep <- rowSums(counts(dds_infected)) > 10
dds_infected <- dds_infected[keep,]
nrow(dds_infected)


## Applying VST transformation
vsd <- vst(dds_infected, blind = FALSE)
# head(assay(vsd), 3)
colData(vsd)
vsd_coldata <- colData(vsd)
dds_infected <- estimateSizeFactors(dds_infected)

### Euclidean Distance between samples

sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$Sample_Name
colnames(sampleDistMatrix) <- vsd$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(EuclideanDistanceHeatmap <- pheatmap(sampleDistMatrix,
                                      clustering_distance_rows = sampleDists,
                                      clustering_distance_cols = sampleDists,
                                      main = "Sample-to-Sample Euclidean Distance of Samples",
                                      col = colors
))

### Poisson Distance between Samples
poisd <- PoissonDistance(t(counts(dds_infected))) # raw counts or unnormalised data
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- dds_infected$Sample_Name
colnames(samplePoisDistMatrix) <- dds_infected$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(poisson_dist_plot <- pheatmap(samplePoisDistMatrix,
                               clustering_distance_rows = poisd$dd,
                               clustering_distance_cols = poisd$dd,
                               main = "Sample-to-Sample Poisson Distance of Samples",
                               col = colors
))
```

```{r}
# PCA Plot

### Functions for Plot aethetics and saving PCA Plots
color_values <- c(
  "red", "red", "red", "red", "red", "red", "red","red", "red", "red", "red",
   "red", "red", "blue", "red", "red","black", "black",  "red", "blue"
)
# the basic set of common aesthetic settings for PCA plots,
theme.my.own <- list(
  theme_bw(),
  geom_point(size = 3),
  coord_fixed(),
  scale_y_continuous(
    breaks = seq(-20, 20, 5),
    sec.axis = sec_axis(~. * 1,
                        labels = NULL,
                        breaks = NULL
    )
  ),
  scale_x_continuous(
    breaks = seq(-20, 20, 5),
    sec.axis = sec_axis(~. * 1,
                        labels = NULL,
                        breaks = NULL
    )
  ),
  theme_classic(),
  geom_hline(yintercept = 0, color = "gray", size = 1),
  geom_vline(xintercept = 0, color = "gray", size = 1),
  theme(
    text = element_text(size = 15),
    axis.text = element_text(size = 15),
    legend.position = "bottom",
    aspect.ratio = 1
  ),
  # geom_text(size = 4, hjust = 0, vjust = 0)
  geom_text_repel(size = 5, min.segment.length = 0.5)
)


## Calculating all PCA Values


plotPCA_local <- function(object,
                          intgroup = "condition",
                          ntop = 500,
                          returnData = TRUE,
                          nPC = 4) {
  # calculate the variance for each gene
  rv <- rowVars(assay(object))
  ntop <- 500
  # select the ntop genes by variance
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))
  # summary(pca)
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <-
    as.data.frame(colData(object)[, intgroup, drop = FALSE])
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = ":"))
  } else {
    colData(object)[[intgroup]]
  }
  # assembly the data for the plot
  d <- cbind(
    pca$x[, seq_len(min(nPC, ncol(pca$x))), drop = FALSE],
    data.frame(group = group, intgroup.df, name = colnames(object))
  )
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:nPC]
    # l <- list(pca,d)
    # return(l)
    return(d)
  }
}


## PCA Plot with VST Data

### Function for calculating percentvar

percentvar_calculation <- function(pcaData_variable) {
  # function to calculate percentvar for different variables
  percentvar_variable <- round(100 * attr(pcaData_variable, "percentVar"), digits = 3)
  return(percentvar_variable)
}

savingFunction <- function(plotname, metadatacolumn) {
  # Function to save the PCA plots
  ggsave(
    filename =
      glue("C:/Users/kesha/Documents/Eva_TotalRNASeq/PCAplot_{metadatacolumn}.png"),
    plot = plotname,
    dpi = 300,
    width = 10,
    height = 10,
    units = "in"
  )
}

pcaData_infected <- plotPCA_local(vsd, intgroup = c("condition", "Sample_Name"), returnData = T)
pcaData_infected
percentVar_infected <- percentvar_calculation(pcaData_infected)
percentVar_infected

(PCAplot_vst <- ggplot(
  pcaData_infected,
  aes(
    x = PC1,
    y = PC2,
    color = Sample_Name,
    label = Sample_Name
  )
) +
  xlab(paste0("PC1: ", percentVar_infected[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar_infected[2], "% variance")) +
  ggtitle("PCA") +
  scale_colour_manual(values = color_values) +
  theme.my.own)
savingFunction(PCAplot_vst, metadatacolumn = "condition")


# calculate the variance for top 500 gene
rv <- rowVars(assay(vsd))
ntop <- 500
# select the ntop genes by variance
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
df1 <- t(assay(vsd)[select,])

res.pca <- PCA(df1, graph = FALSE, scale.unit = FALSE)
summary.PCA(res.pca)

# Visualize eigenvalues/variances
fviz_screeplot(res.pca, addlabels = TRUE)

eig.val <- get_eigenvalue(res.pca)
eig.val


```

```{r fig.width=8}
var <- get_pca_var(res.pca)
# fviz_pca_var(res.pca, repel = TRUE)
# 
head(var$cos2,25)
# library("corrplot")
# corrplot(var$cos2, is.corr = TRUE, diag = TRUE, )

```

```{r}
## Genes + PCA Biplots

fviz_pca_biplot(res.pca,
                repel = TRUE,
                #gradient.cols = c("gray","pink", "blue", "yellow", "green", "red", "black")
)

heat.colors <- brewer.pal(6, "RdYlBu")
fviz_pca_var(res.pca,
             col.var = "contrib", repel = TRUE,
             gradient.cols = c("Gray", "blue", "pink","yellow", "orange", "green", "red", "black"),
)

# Contributions of variables to PC2
fviz_contrib(res.pca, choice = "var", axes = 2, top = 25)
#
# # Contributions of variables to PC1
fviz_contrib(res.pca, choice = "var", axes = 1, top = 25)

```
```{r}
## Hierarchical Clustering

### applying rlog Transformation
rld <- rlog(dds_infected, blind = FALSE)
head(assay(rld), 3)

### Extract the rlog matrix from the object
rld_mat <- assay(rld) # assay() is function from the "SummarizedExperiment" package that was loaded when you loaded DESeq2
### Compute pairwise correlation values
rld_cor <- cor(rld_mat) ## cor() is a base R function
head(rld_cor) ## check the output of cor(), make note of the rownames and colnames

### Plot heatmap
heat.colors <- brewer.pal(6, "RdYlBu")
(Hclust_plot <- pheatmap(rld_cor,
                         color = heat.colors,
                         main = "Heirarchical Clustering of Samples - Correlation Matrix"
                         # filename = 'C:/Users/kesha/Documents/Eva_TotalRNASeq/Hclust_plot.tiff'
))
# Hclust_plot
```

```{r}
# DGE Results

### Running the differential expression pipeline
dds1_infected <- DESeq(dds_infected)
# str(dds1)
### Building the results table
res_infected <- results(dds1_infected,
                        contrast = c("condition", "adult", "d7")
)

resdf_infected <- as.data.frame(res_infected) # convert the results table to a df
head(resdf_infected, 20)

write.csv(resdf_infected, file = "C:/Users/kesha/Documents/Eva_TotalRNASeq/results_DGE.csv")
```
## MA Plot

```{r}
plotMA_res_infected <- plotMA(res_infected, ylim = c(-2, 2))
```

### Histogram of p-values

```{r}
hist(res_infected$pvalue, breaks = 50, col = "grey50", border = "blue")
```
```{r}
head(resdf_infected)
```
```{r}
resdf_infected <- tibble::rownames_to_column(resdf_infected, "symbol")
gn <- resdf_infected$symbol
# Mapping the Symbol to ENTREZ ID
entrez <- mapIds(
  org.Mm.eg.db,
  keys = gn,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)
ensembl_id <- mapIds(
  org.Mm.eg.db,
  keys = gn,
  column = "ENSEMBL",
  keytype = "SYMBOL",
  multiVals = "first"
)
resdf_infected$entrez <- entrez
resdf_infected$ensemblID <- ensembl_id
```

Omit NA values from symbol and respective rows!

```{r}
resdf_infected <- resdf_infected %>% filter(!is.na(symbol) & !is.na(entrez))
nrow(resdf_infected)
```


## Effect of Transformations on Variance

-   These set of plots depict the standard deviation of transformed data
    (across samples), against mean.

### Based on Shifted Log Transformation

```{r}
ntd <- normTransform(dds_infected)
meanSdPlot(assay(ntd))
```

### Based on Rlog Transformation

```{r}
meanSdPlot(assay(rld))
```

### Based on Variance Stabilizing Transformation

```{r}
meanSdPlot(assay(vsd))
```
# Volcano Plots

## Volcano Plots based on Enhanced Volcano

```{r fig.height= 10, fig.width= 10}
p1 <- EnhancedVolcano(resdf_infected,
  lab = resdf_infected$symbol,
  x = "log2FoldChange",
  y = "pvalue",
  xlab = bquote(~ Log[2] ~ "FoldChange"),
  pCutoff = 0.05,
  FCcutoff = 1.0,
  title = "DE genes: Log2FoldChange Vs -Log10 Pvalue",
  subtitle = bquote(~ Log[2] ~ "|FoldChange| = 1, pvalue < 0.05"),
  pointSize = 2.0,
  labSize = 5.5,
  boxedLabels = FALSE,
  gridlines.major = FALSE,
  gridlines.minor = FALSE,
  colAlpha = 0.5,
  legendPosition = "bottom",
  legendLabSize = 12,
  legendIconSize = 4.0,
  drawConnectors = T,
  widthConnectors = 0.75,
  max.overlaps = 8,
  axisLabSize = 22
)
(volcano1 <- p1 + scale_y_continuous(
  limits = c(0, 8),
  breaks = seq(0, 8, 1),
  sec.axis = sec_axis(~ . * 1,
    labels = NULL,
    breaks = NULL
  )
) +
  scale_x_continuous(
    limits = c(-4, 5),
    breaks = seq(-4, 5, 1),
    sec.axis = sec_axis(~ . * 1,
      labels = NULL,
      breaks = NULL
    )
  ) +
  xlab(expression(DownRegulated %<->% UpRegulated))
)
ggsave(
  filename = "C:/Users/kesha/Documents/Eva_TotalRNASeq/Volcano1_L2fcVsPvalue.png",
  plot = volcano1,
  dpi = 300,
  width = 5,
  height = 5,
  units = "in"
)
```
## volcano Plot of Log2FC vs Padj

```{r fig.height= 10, fig.width= 10}
p2 <- EnhancedVolcano(resdf_infected,
  lab = resdf_infected$symbol,
  x = "log2FoldChange",
  y = "padj",
  xlab = bquote(~ Log[2] ~ "FoldChange"),
  ylab = bquote(~ Log[10] ~ "Padj"),
  pCutoff = 0.05,
  FCcutoff = 1.0,
  title = "DE genes: Log2FoldChange Vs -Log10 Padj",
  subtitle = bquote(~ Log[2] ~ "|FoldChange| = 1, pvalue < 0.05"),
  pointSize = 2.0,
  labSize = 5.0,
  boxedLabels = FALSE,
  gridlines.major = FALSE,
  gridlines.minor = FALSE,
  colAlpha = 0.5,
  legendPosition = "bottom",
  legendLabSize = 12,
  legendIconSize = 4.0,
  drawConnectors = T,
  widthConnectors = 0.75,
  max.overlaps = 8,
  axisLabSize = 22
)
(volcano2 <- p2 + scale_y_continuous(
  limits = c(0, 4),
  breaks = seq(0, 4, 1),
  sec.axis = sec_axis(~ . * 1,
    labels = NULL,
    breaks = NULL
  )
) +
  scale_x_continuous(
    limits = c(-3, 4),
    breaks = seq(-3, 4, 1),
    sec.axis = sec_axis(~ . * 1,
      labels = NULL,
      breaks = NULL
    )
  ) +
  xlab(expression(DownRegulated %<->% UpRegulated))
)
ggsave(
  filename = "C:/Users/kesha/Documents/Eva_TotalRNASeq/Volcano2_L2fcVsPadj.png",
  plot = volcano2,
  dpi = 300,
  width = 10,
  height = 10,
  units = "in"
)
```

## Significant Differentially Expressed Genes

```{r}
resdf_infected$diffexpressed <- "NS"
# if log2Foldchange > 1.0 and pvalue < 0.05, set as "UP"
resdf_infected$diffexpressed[resdf_infected$log2FoldChange > 1.0 & resdf_infected$pvalue < 0.05] <- "UP"
# if log2Foldchange < -1.0 and pvalue < 0.05, set as "DOWN"
resdf_infected$diffexpressed[resdf_infected$log2FoldChange < -1.0 & resdf_infected$pvalue < 0.05] <- "DOWN"
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
resdf_infected$delabel <- NA
resdf_infected$delabel[resdf_infected$diffexpressed != "NS"] <- resdf_infected$symbol[resdf_infected$diffexpressed != "NS"]
```

### Arrive at relevant genes by imposing thresholds.

```{r}
sigsdf_infected <- resdf_infected[(abs(resdf_infected$log2FoldChange) > 1) & (resdf_infected$pvalue < 0.05), ]
nrow(sigsdf_infected)
```
### Gender Genes Removed from Table

```{r}
sigsdf_infected <- filter(
  sigsdf_infected,
  symbol != "Xist",
  symbol != "Jpx",
  symbol != "Ftx",
  symbol != "Tsx",
  symbol != "Cnbp2"
)
nrow(sigsdf_infected)
# head(sigsdf_infected)
```

Therefore, gender genes werent so much in play in terms of significance!

```{r}
write.csv(sigsdf_infected, "C:/Users/kesha/Documents/Eva_TotalRNASeq/SignificantGenes_IvC.csv")
```
### Significant Differentialy Expressed Genes (SDEG)

```{r}
head(sigsdf_infected)
```


### Number of Genes from different strains that are contributing to UP/DOWN regulation.

```{r}
sigsdf_infected_UP <- sigsdf_infected[(sigsdf_infected$log2FoldChange) > 1, ] # UP Regulation Table
sigsdf_infected_DOWN <- sigsdf_infected[(sigsdf_infected$log2FoldChange) < -1, ] # DOWN Regulation Table
nrow(sigsdf_infected_UP)
nrow(sigsdf_infected_DOWN)
```



```{r}
# sorted based on highest Log2FC value
UpGene_infected <- sigsdf_infected_UP[order(-sigsdf_infected_UP$log2FoldChange, decreasing = TRUE), ]$symbol
DownGene_infected <- sigsdf_infected_DOWN[order(-sigsdf_infected_DOWN$log2FoldChange, decreasing = TRUE), ]$symbol
```
```{r}
write.csv(
  UpGene_infected,
  "C:/Users/kesha/Documents/Eva_TotalRNASeq/UpGenelist.csv"
)
write.csv(
  DownGene_infected,
  "C:/Users/kesha/Documents/Eva_TotalRNASeq/DownGenelist.csv"
)
```

### Top20 UP Genes and Top20 DOWN Genes

#### UpGene_infected
```{r}
head(UpGene_infected, 20)
```

#### DownGene_infected
```{r}
head(DownGene_infected, 20)
```

```{r}
# Determining the significant Genes based on Log2FC and pvalue thresholds
sigs2df_infected <- resdf_infected[(abs(resdf_infected$log2FoldChange) > 1) & (resdf_infected$pvalue < 0.05), ]
```

## Z-score based Gene Heatmaps

### with Whole table (all genes together!)

```{r}
# mat <- counts(dds2_infected, normalized = TRUE)[rownames(sigsdf_infected),]
mat <- counts(dds_infected, normalized = TRUE)[(sigsdf_infected$symbol), ]
mat.zs <- t(apply(mat, 1, scale)) # Calculating the zscore for each row
colnames(mat.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat.zs)
```

```{r}
mat2 <- counts(dds_infected, normalized = TRUE)[(sigsdf_infected$symbol), ]
mat2.zs <- t(apply(mat2, 1, scale)) # Calculating the zscore for each row
colnames(mat2.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat2.zs)
# pheatmap(mat2.zs, cluster_cols = TRUE, cluster_rows = FALSE, show_rownames = FALSE)
```
```{r}
newHP <- Heatmap(mat2.zs,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  column_labels = colnames(mat2.zs),
  name = "DE Genes Heatmap",
  show_row_names = FALSE,
  use_raster = TRUE,
  raster_quality = 5,
  column_names_gp = grid::gpar(fontsize = 12),
  # width = unit(8, "in"),
  # height = unit(8, "in")
  row_labels = sigs2df_infected[rownames(mat2.zs), ]$symbol
)
newHP
```
```{r fig.height= 500, fig.width= 10}
# LongHeatMap_Allgenes <- Heatmap(mat2.zs,
#   cluster_columns = TRUE,
#   cluster_rows = TRUE,
#   column_labels = colnames(mat2.zs),
#   row_labels = sigsdf_infected$symbol,
#   name = "All DE Genes",
#   show_row_names = TRUE,
#   use_raster = TRUE,
#   raster_quality = 5,
#   column_names_gp = grid::gpar(fontsize = 12),
# )
# LongHeatMap_Allgenes
# saveplot(LongHeatMap_Allgenes, "LongHeatMap_Allgenes")
```

Need to filter these results to accommodate better the heat maps and also volcano plots

### with tighter constraints (all genes together!)

```{r}
sigs1df_infected <- resdf_infected[(resdf_infected$baseMean > 15000) & (abs(resdf_infected$log2FoldChange) >  2) & (resdf_infected$pvalue < 0.05), ]
mat1 <- counts(dds_infected, normalized = TRUE)[(sigs1df_infected$symbol), ]
mat1.zs <- t(apply(mat1, 1, scale)) # Calculating the zscore for each row
colnames(mat1.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat1.zs)
```

```{r fig.height=12}
Heatmap(mat1.zs,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  column_labels = colnames(mat1.zs),
  name = "DE Genes",
  row_labels = sigs1df_infected$symbol,
  column_names_gp = grid::gpar(fontsize = 18),
  row_names_gp = grid::gpar(fontsize = 20)
)
```
# GO Terms with clusterProfiler

## GO Terms for UP Regulated Genes

### GO over-representation analysis for UP Regulated Genes
```{r}
UPgene_ENS_ID <- (sigsdf_infected_UP$ensemblID)
GO_UPRegResults_infected <- enrichGO(
  gene = UPgene_ENS_ID,
  OrgDb = "org.Mm.eg.db",
  keyType = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)
```

```{r}
GO_UpRegdf_infected <- as.data.frame(GO_UPRegResults_infected)
```
```{r}
head(GO_UpRegdf_infected, 20)
```

```{r}
write.csv(GO_UpRegdf_infected, "C:/Users/kesha/Documents/Eva_TotalRNASeq/GO_UpReg_results.csv")
```

```{r fig.height=12}
GO_UPReg_Barplot_infected <- plot(barplot(GO_UPRegResults_infected, 
                                          showCategory = 20, 
                                          font.size = 25,
                                          title = "InfectedVsControl Up Regulated",
                                          #label_format = 40
                                          ))
saveplot(GO_UPReg_Barplot_infected, "GO_UPReg_Barplot_infected")
```

```{r fig.height=10}
GO_UPReg_Dotplot_infected <- plot(dotplot(GO_UPRegResults_infected, 
                                          showCategory = 20, 
                                          font.size = 30,
                                          title = "InfectedVsControl Up Regulated",
                                          #label_format = 60
                                          ))
#GO_UPReg_Dotplot_infected1 <- GO_UPReg_Dotplot_infected + scale_y_discrete(labels=function(y) str_wrap(y, width=40))
saveplot(GO_UPReg_Dotplot_infected, "GO_UPReg_Dotplot_infected")
```

```{r fig.height=10, fig.width=12}
GO_UPReg_Cnetplot_infected <- plot(cnetplot(GO_UPRegResults_infected, 
                                            showCategory = 13, 
                                          font.size = 25,
                                          label_format = 75
                                          ))
saveplot(GO_UPReg_Cnetplot_infected, "GO_UPReg_Cnetplot_infected")
```

### Upset Plot

```{r fig.height= 8, fig.width=12}
GO_UPReg_Upsetplot <- plot( upsetplot(GO_UPRegResults_infected))
saveplot(GO_UPReg_Upsetplot, "GO_UPReg_Upsetplot")
```

### Heatplot

The heatplot is similar to cnetplot, while displaying the relationships as a heatmap. The gene-concept network may become too complicated if user want to show a large number significant terms. The heatplot can simplify the result and more easy to identify expression patterns.

```{r fig.height= 12, fig.width=15}
GO_UPReg_Heatplot <- plot(heatplot(GO_UPRegResults_infected, label_format = 150))
saveplot(GO_UPReg_Heatplot, "GO_UPReg_Heatplot")
```

### Tree Plot of Enriched Terms

```{r fig.height= 10, fig.width=15}
edox2 <- pairwise_termsim(GO_UPRegResults_infected)
GO_UPReg_enrichtreeplot <- plot(treeplot(edox2))
saveplot(GO_UPReg_enrichtreeplot, "GO_UPReg_enrichtreeplot")
```

## GO Terms for Down Regulated Genes

### GO over-representation analysis for DOWN Regulated Genes

```{r}
DOWNgene_ENS_ID <- (sigsdf_infected_DOWN$ensemblID)
GO_DOWNRegResults_infected <- enrichGO(
  gene = DOWNgene_ENS_ID,
  OrgDb = "org.Mm.eg.db",
  keyType = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)
GO_DOWNRegdf_infected <- as.data.frame(GO_DOWNRegResults_infected)
head(GO_DOWNRegdf_infected)
```

```{r}
write.csv(GO_DOWNRegdf_infected,
          "C:/Users/kesha/Documents/Eva_TotalRNASeq/GO_DOWNReg_results.csv")
```

```{r}
GO_DOWNReg_Barplot_infected <- plot(barplot(GO_DOWNRegResults_infected))
saveplot(GO_DOWNReg_Barplot_infected, "GO_DOWNReg_Barplot_infected")
```

```{r}
GO_DOWNReg_Dotplot_infected <- plot(dotplot(GO_DOWNRegResults_infected))

saveplot(GO_DOWNReg_Dotplot_infected, "GO_DOWNReg_Dotplot_infected")
```

```{r }
GO_DOWNReg_Cnetplot_infected <- plot(cnetplot(GO_DOWNRegResults_infected,
                                                showCategory = 13,
                                                circular = FALSE,
                                                #colorEdge = TRUE
                                                ))

saveplot(GO_DOWNReg_Cnetplot_infected, "GO_DOWNReg_Cnetplot_infected")
```

### Upset Plot

The upsetplot is an alternative to cnetplot for visualizing the complex
association between genes and gene sets. It emphasizes the gene
overlapping among different gene sets.

```{r}
GO_DOWNReg_Upsetplot <- plot( upsetplot(GO_DOWNRegResults_infected))
saveplot(GO_DOWNReg_Upsetplot, "GO_DOWNReg_Upsetplot")
```

### Heat Plot

```{r}
GO_DOWNReg_Heatplot <- plot(heatplot(GO_DOWNRegResults_infected))
saveplot(GO_DOWNReg_Heatplot, "GO_DOWNReg_Heatplot")
```

### Tree Plot of Enriched Terms

```{r}
edox1 <- pairwise_termsim(GO_DOWNRegResults_infected)
GO_DOWNReg_enrichtreeplot <- plot(treeplot(edox1))
saveplot(GO_DOWNReg_enrichtreeplot, "GO_DOWNReg_enrichtreeplot")
```